### PR TITLE
Add code cache to provider service

### PR DESCRIPTION
**Pull request by Sajma27**
# Summary
A description of the changes proposed in the pull request.

Fixes #number

## Checklist
- [ ] Change is small and easy to review (please split big changes into multiple PRs)
- [ ] Change is consistent with architecture
- [ ] Change is consistent with test architecture
- [ ] Change is consistent with naming conventions
- [ ] New code is covered with tests
- [ ] Tests related to old code are updated
- [ ] Documentation is up to date with changes

### CODE REVIEW

File: universal-login-commons/src/integration/ethereum/ProviderService.ts [hunk 1]
```diff
@@ -4,11 +4,19 @@ import {fetchHardforkVersion} from './fetchHardforkVersion';
 import {NetworkVersion} from '../../core/utils/messages/computeGasData';
 
 export class ProviderService {
+  private cachedContractCodes: Map<string, Promise<string>>;
```
Comment (JustynaBroniszewska) [inline universal-login-commons/src/integration/ethereum/ProviderService.ts:7]: You can make it in a similar way as there: WalletContractService

Comment (Sajma27) [reply universal-login-commons/src/integration/ethereum/ProviderService.ts]: Done

File: universal-login-commons/src/integration/ethereum/ProviderService.ts [hunk 2]
```diff
@@ -4,11 +4,19 @@ import {fetchHardforkVersion} from './fetchHardforkVersion';
...
   constructor(private provider: providers.Provider) {
   }
 
   getCode(contractAddress: string) {
-    return this.provider.getCode(contractAddress);
+    return this.cachedContractCodes[contractAddress] || this.getCodeFromProviderAndCacheIt(contractAddress);
+  }
+
+  private getCodeFromProviderAndCacheIt(contractAddress: string) {
+    const code = this.provider.getCode(contractAddress);
```
Comment (JustynaBroniszewska) [inline universal-login-commons/src/integration/ethereum/ProviderService.ts:17]: getCode is async, isn't it?

Comment (Sajma27) [reply universal-login-commons/src/integration/ethereum/ProviderService.ts:17]: Yeah, but we are storing and returning Promise<string>

Comment (Sajma27) [reply universal-login-commons/src/integration/ethereum/ProviderService.ts:17]: Okej, done

Comment (JustynaBroniszewska) [reply universal-login-commons/src/integration/ethereum/ProviderService.ts:17]: Use await here

**Pull request merged**